### PR TITLE
if_perl.xs: fix compatibility for Perl 5.36 and up

### LINKED ISSUE #12543: vim not build with upcoming Perl 5.38
### Steps to reproduce

I am preparing update to Perl 5.38 for Fedora 39. During the preparation I tried to rebuild all dependencies with it.
I found that build of vim was failing with new Perl.
For the tests I used the latest devel release [Perl 5.37.11](https://cpan.metacpan.org/authors/id/S/SH/SHAY/perl-5.37.11.tar.gz) 

### Expected behaviour

Successful build of vim

### Version of Vim

9.0 1627

### Environment

Fedora 39 (rawhide)
Devel release of Perl  5.37.11

### Logs and stack traces

```shell
Complete log coulg be found on <https://jplesnik.fedorapeople.org/5.38/vim-build.log>
Examples from log: 

if_perl.xs: In function 'Perl_SvTRUE_common': 
/usr/lib64/perl5/CORE/embed.h:109:49: warning: implicit declaration of function 'Perl_SvPVXtrue'; did you mean 'Perl_sv_true'? [-Wimplicit-function-declaration]
  109 | # define SvPVXtrue(a)                           Perl_SvPVXtrue(aTHX_ a)
      |                                                 ^~~~~~~~~~~~~~
if_perl.xs:724:16: note: in expansion of macro 'SvPVXtrue'
  724 |         return SvPVXtrue(sv);
      |                ^~~~~~~~~
if_perl.xs: In function 'Perl_SvTRUE':
<...>
/usr/bin/ld: /tmp/ccxXLZZi.ltrans63.ltrans.o: in function `ex_perl':
/builddir/build/BUILD/vim90/src/if_perl.xs:745: undefined reference to `Perl_SvGETMAGIC'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1108: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:724: undefined reference to `Perl_SvPVXtrue'
/usr/bin/ld: /tmp/ccxXLZZi.ltrans63.ltrans.o: in function `perl_to_vim.lto_priv.0':
/builddir/build/BUILD/vim90/src/if_perl.xs:1210: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1323: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1199: undefined reference to `Perl_SvNV'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1203: undefined reference to `Perl_SvIV'
/usr/bin/ld: /tmp/ccxXLZZi.ltrans64.ltrans.o: in function `XS_VIM_DoCommand':
/builddir/build/BUILD/vim90/src/if_perl.c:1789: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /tmp/ccxXLZZi.ltrans64.ltrans.o: in function `ex_perldo':
/builddir/build/BUILD/vim90/src/if_perl.xs:1433: undefined reference to `Perl_SvPV_helper'

```

### LINKED ISSUE #21164: vim not able to build with Perl >=5.37
**Description**
Not possible to build vim with **perl >= 5.37**

```
Complete log coulg be found on <https://jplesnik.fedorapeople.org/5.38/vim-build.log>
Examples from log: 

if_perl.xs: In function 'Perl_SvTRUE_common': 
/usr/lib64/perl5/CORE/embed.h:109:49: warning: implicit declaration of function 'Perl_SvPVXtrue'; did you mean 'Perl_sv_true'? [-Wimplicit-function-declaration]
  109 | # define SvPVXtrue(a)                           Perl_SvPVXtrue(aTHX_ a)
      |                                                 ^~~~~~~~~~~~~~
if_perl.xs:724:16: note: in expansion of macro 'SvPVXtrue'
  724 |         return SvPVXtrue(sv);
      |                ^~~~~~~~~
if_perl.xs: In function 'Perl_SvTRUE':
<...>
/usr/bin/ld: /tmp/ccxXLZZi.ltrans63.ltrans.o: in function `ex_perl':
/builddir/build/BUILD/vim90/src/if_perl.xs:745: undefined reference to `Perl_SvGETMAGIC'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1108: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:724: undefined reference to `Perl_SvPVXtrue'
/usr/bin/ld: /tmp/ccxXLZZi.ltrans63.ltrans.o: in function `perl_to_vim.lto_priv.0':
/builddir/build/BUILD/vim90/src/if_perl.xs:1210: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1323: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1199: undefined reference to `Perl_SvNV'
/usr/bin/ld: /builddir/build/BUILD/vim90/src/if_perl.xs:1203: undefined reference to `Perl_SvIV'
/usr/bin/ld: /tmp/ccxXLZZi.ltrans64.ltrans.o: in function `XS_VIM_DoCommand':
/builddir/build/BUILD/vim90/src/if_perl.c:1789: undefined reference to `Perl_SvPV_helper'
/usr/bin/ld: /tmp/ccxXLZZi.ltrans64.ltrans.o: in function `ex_perldo':
/builddir/build/BUILD/vim90/src/if_perl.xs:1433: undefined reference to `Perl_SvPV_helper'
```

**Steps to Reproduce**
```
wget -O perl.tar.gz https://cpan.metacpan.org/authors/id/S/SH/SHAY/perl-5.37.11.tar.gz
tar -xaf perl.tar.gz && cd perl-5.37.11
<install make and gcc>
sh Configure
make && sudo make install

git clone https://github.com/vim/vim
cd vim
./configure --enable-gui=no --with-features=Huge --enable-perlinterp=dynamic
make
```

**Expected behavior**
no build errors by Perl

**Perl configuration**
```
# perl -V 5.37.11
```

### PR DESCRIPTION
**Pull request by pheiduck**
Test Env:
Fedora 39 + Perl 5.37.11 and Perl 5.38.0-RC1
`./configure --enable-gui=no --with-features=Huge --enable-perlinterp`

Fixes: https://github.com/vim/vim/issues/12543
Closes: https://github.com/Perl/perl5/issues/21164

### CODE REVIEW

File: src/if_perl.xs [hunk 1]
```diff
@@ -56,6 +56,8 @@
 #ifdef __GNUC__
 # pragma GCC diagnostic push
 # pragma GCC diagnostic ignored "-Wunused-variable"
+# pragma GCC diagnostic ignored "-Wincompatible-pointer-types"
+# pragma GCC diagnostic ignored "-Wshadow"
```
Comment (ichizok) [inline src/if_perl.xs]: I think should not ignore these warnings (`-Wincompatible-pointer-types` and `-Wshadow`) by pragma.
For CI, should add them to `PERL_CFLAGS_EXTRA` in `ci/config.mk.sed`.

File: src/if_perl.xs [hunk 2]
```diff
@@ -40,7 +40,7 @@
 /* Work around for perl-5.18.
  * Don't include "perl\lib\CORE\inline.h" for now,
  * include it after Perl_sv_free2 is defined. */
-#ifdef DYNAMIC_PERL
+#if (PERL_REVISION == 5) && (PERL_VERSION == 18)
```
Comment (k-takata) [inline src/if_perl.xs]: This comment says "work around for perl-5.18", but I think actually this is a workaround for perl-5.18 and later.
Not sure which versions are relevant, though.

File: src/if_perl.xs (hunk 1 above)
Comment (pheiduck) [reply src/if_perl.xs]: done: https://github.com/vim/vim/pull/12575/commits/6abf8cdab0bf03ea05cec19d538d4387d730a933

File: src/if_perl.xs (hunk 2 above)
Comment (pheiduck) [reply src/if_perl.xs]: done: https://github.com/vim/vim/pull/12575/commits/6abf8cdab0bf03ea05cec19d538d4387d730a933

**Comment by pheiduck** [comment]
macOS huge tests failures:
https://github.com/vim/vim/actions/runs/5354406792/jobs/9711406056#step:8:15020
not sure how to solve them.

**Comment by chrisbra** [comment]
> macOS huge tests failures:
https://github.com/vim/vim/actions/runs/5354406792/jobs/9711406056#step:8:15020
not sure how to solve them.

That seems unrelated, happens currently also on master: https://github.com/vim/vim/actions/runs/5350728843/jobs/9703711955
and seems more related to patch 1aa5f1c21fea60fe64cdcdf6990bf43ea5cc78a0

**Comment by brammool** [comment]
The Mac failures are fixed now.  Please sync to head.

**Comment by pheiduck** [comment]
@brammool done.

**Review by ichizok — APPROVED** [review]

**Comment by pheiduck** [comment]
@brammool any reason why this is hold back?

**Comment by zdohnal** [comment]
I saw several warnings about incompatible argument types when I applied the patch and the following patch fixed them:

```
@@ -402,14 +402,14 @@ static bool (*Perl_sv_2bool)(pTHX_ SV*);
 static IV (*Perl_sv_2iv)(pTHX_ SV*);
 static SV* (*Perl_sv_2mortal)(pTHX_ SV*);
 # if (PERL_REVISION == 5) && (PERL_VERSION >= 8)
-static char* (*Perl_sv_2pv_flags)(pTHX_ SV*, STRLEN*, I32);
+static char* (*Perl_sv_2pv_flags)(pTHX_ SV*, STRLEN* const, const U32);
 static char* (*Perl_sv_2pv_nolen)(pTHX_ SV*);
 # else
 static char* (*Perl_sv_2pv)(pTHX_ SV*, STRLEN*);
 # endif
 static char* (*Perl_sv_2pvbyte)(pTHX_ SV*, STRLEN*);
 # if (PERL_REVISION == 5) && (PERL_VERSION >= 32) 
-static char* (*Perl_sv_2pvbyte_flags)(pTHX_ SV*, STRLEN*, I32);
+static char* (*Perl_sv_2pvbyte_flags)(pTHX_ SV*, STRLEN* const, const U32);
 # endif
 static SV* (*Perl_sv_bless)(pTHX_ SV*, HV*);
 # if (PERL_REVISION == 5) && (PERL_VERSION >= 8)
```

otherwise the patch in the PR looks good.

**Comment by pheiduck** [comment]
There are some deprecated declarations on **Perl 5.39.0**
so ignoring the `-Wdeprecated-declarations` will hide then build w/ **Perl 5.39**

**Comment by pheiduck** [comment]
```
if_perl.xs: In function ‘XS_VIM_Buffers’:
if_perl.xs:1652:9: warning: ‘Perl_dowantarray’ is deprecated [-Wdeprecated-declarations]
 1652 |         if (GIMME == G_SCALAR)
      |         ^~
In file included from /usr/local/lib/perl5/5.39.0/x86_64-linux/CORE/perl.h:6188,
                 from if_perl.xs:62:
/usr/local/lib/perl5/5.39.0/x86_64-linux/CORE/proto.h:1019:1: note: declared here
 1019 | Perl_dowantarray(pTHX)
      | ^~~~~~~~~~~~~~~~
if_perl.xs: In function ‘XS_VIM_Windows’:
if_perl.xs:1703:9: warning: ‘Perl_dowantarray’ is deprecated [-Wdeprecated-declarations]
 1703 |         if (GIMME == G_SCALAR)
      |         ^~
/usr/local/lib/perl5/5.39.0/x86_64-linux/CORE/proto.h:1019:1: note: declared here
 1019 | Perl_dowantarray(pTHX)
```

Doesn't interrupted the build.

**Comment by pheiduck** [comment]
@brammool could please revise this PR again?
Thank You!

**Comment by brammool** [comment]
> @brammool any reason why this is hold back?

Not specificly.  My inbox is rather full, I'm trying to include PRs and
suggested improvements in a "most important first" order.  The more
patches I have waiting the slower this gets...

-- 
FIRST HEAD:  Oh! quick! get the sword out I want to cut his head off.
THIRD HEAD:  Oh, cut your own head off.
SECOND HEAD: Yes - do us all a favour.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by pheiduck** [comment]
@chrisbra It would be very cool if we can merge this one.
So Distro Maintainers do not have to apply this as a separate patch. :)

**Pull request closed without merging**